staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -697,7 +697,7 @@ func (r *Reflector) watchList(stopCh <-chan struct{}) (watch.Interface, error) {
 	// we utilize the temporaryStore to ensure independence from the current store implementation.
 	// as of today, the store is implemented as a queue and will be drained by the higher-level
 	// component as soon as it finishes replacing the content.
-	checkWatchListConsistencyIfRequested(stopCh, r.name, resourceVersion, r.listerWatcher, temporaryStore)
+	checkWatchListConsistencyIfRequested(wait.ContextForChannel(stopCh), r.name, resourceVersion, wrapListFuncWithContext(r.listerWatcher.List), temporaryStore.List)
```

Comment:
I don't understand why you want're switching the signature here - the switch to context makes sense, but the other two changes, I'm not following and I suggest reverting those.
Both Lister and Store are well defined external interfaces and it seems fine to use them.

Reply:
The `Lister` and `Store` interfaces are convenient for a reflector but not for a generated client. For example, at some point we will extend the generated client to use the `WatchList` method and for consistency checking we will use `checkWatchListConsistencyIfRequested`. If we don't change the signature to accept simple functions then we will have to create types that conform to the interfaces.


The generated client doesn't know `Lister` and `Store` interfaces, for example:
https://github.com/kubernetes/kubernetes/blob/7a37df6d1ab8926a95574f0aab3a9978d20d4e35/staging/src/k8s.io/client-go/kubernetes/typed/apps/v1/deployment.go#L93

The `Store` interface is huge - https://github.com/kubernetes/kubernetes/blob/ec9515a828ec3907ffc66913947adf4e8ee7b9d6/staging/src/k8s.io/client-go/tools/cache/store.go#L40

For the `Lister` interface we would have to create a type that exposed the `List` method.

Having simple functions in the `checkWatchListConsistencyIfRequested` simplifies usage of this function by different components.

Reply:
So you want to enable it also for regular list requests (when we switch them to using streaming lists)?

I started wondering where exactly we should integrate it. I don't think integrating it with generated client is the right option.
I would much rather prefer to integrate it with the changes that you're working as of now in:
https://github.com/kubernetes/kubernetes/pull/122657

So WatchList() call underneath is performing consistency check [if enabled].

In general, rather than doing refactorings now, I would like to see the eventual structure now to see what's the best way of achieving it.

Reply:
The end goal is to modify the generated client's `List` method to use streaming when the `FG` is set. Integrating the detector with the generated client will allow us to simply call the `List` method, which will handle the heavy lifting for us, such as parameter encoding, setting timeouts, and decoding.

The detector will only be enabled in the CI and it is purely for validation purposes.

Reply:
OK - at the high level I know that.
What I'm saying is that I wouldn't add detector in generated client - rather instead integrated it in newly added WatchList() method for restclient.
[And that will affect what/how we want to refactor]

Reply:
Yes, it would be better to add the detector to the WatchList method, but since that would be problematic, I wanted to add the detector to the layer above.

In my opinion, adding the detector to the higher layer is not incorrect.
The primary goal is to change the List method and test whether it works.
Testing will only take place in CI.
The vast majority of users and code use client-go, not the rest client.